core/trino-main/src/main/java/io/trino/sql/gen/columnar/BloomFilter.java

```diff
@@ -0,0 +1,192 @@
+            int selectedPositionsCount = 0;
+            if (block.mayHaveNull()) {
+                for (int index = offset; index < offset + size; index++) {
+                    int position = activePositions[index];
+                    boolean result = block.isNull(position) ? isNullAllowed : filter.test(block, position);
+                    outputPositions[selectedPositionsCount] = position;
+                    selectedPositionsCount += result ? 1 : 0;
+                }
+                return selectedPositionsCount;
+            }
+
+            for (int index = offset; index < offset + size; index++) {
+                int position = activePositions[index];
+                outputPositions[selectedPositionsCount] = position;
+                selectedPositionsCount += filter.test(block, position) ? 1 : 0;
+            }
+            return selectedPositionsCount;
+        }
+
+        @Override
+        public InputChannels getInputChannels()
+        {
+            return inputChannels;
+        }
+    }
+
+    public static final class SliceBloomFilter
+    {
+        private final long[] bloom;
+        private final int bloomSizeMask;
+
+        /**
+         * A Bloom filter for a set of Slice values.
+         * This is approx 2X faster than the Bloom filter implementations in ORC and parquet because
+         * it uses single hash function and uses that to set 3 bits within a 64 bit word.
+         * The memory footprint is up to (4 * values.size()) bytes, which is much smaller than maintaining a hash set of strings.
+         *
+         * @param values List of values used for filtering
+         */
+        public SliceBloomFilter(List<Slice> values)
+        {
+            int bloomSize = getBloomFilterSize(values.size());
+            bloom = new long[bloomSize];
+            bloomSizeMask = bloomSize - 1;
+            for (Slice value : values) {
+                long hashCode = XxHash64.hash(value);
+                // Set 3 bits in a 64 bit word
+                bloom[bloomIndex(hashCode)] |= bloomMask(hashCode);
+            }
+        }
+
+        private static int getBloomFilterSize(int valuesCount)
+        {
+            // Linear hash table size is the highest power of two less than or equal to number of values * 4. This means that the
+            // table is under half full, e.g. 127 elements gets 256 slots.
+            int hashTableSize = Integer.highestOneBit(valuesCount * 4);
+            // We will allocate 8 bits in the bloom filter for every slot in a comparable hash table.
+            // The bloomSize is a count of longs, hence / 8.
+            return Math.max(1, hashTableSize / 8);
+        }
+
+        public boolean test(VariableWidthBlock block, int position)
+        {
+            return contains(block.getRawSlice(), block.getRawSliceOffset(position), block.getSliceLength(position));
+        }
+
+        @VisibleForTesting
+        public boolean contains(Slice data)
```

Comment:
This method is only used in test. Can we use contains(data, offset, length) in test instead?